### ISSUE TITLE
[check_free_space] ERROR

### ISSUE THREAD

**Issue by itsbradwhite** [grahampugh/erase-install#324]
![Screenshot 2022-12-15 at 4 10 04 PM](https://user-images.githubusercontent.com/87879319/207992832-51ca653b-5cce-4f2f-af90-fadeac5c13ce.png)
**Describe the bug**

We just unblocked Ventura in my org. It looks like erase-install is still the most recommended approach for allowing standard-users to run major updates. I've had much success with this script in the past.

I'm seeing  [check_free_space] ERROR's.

This particular computer, the user ran the policy three times. It doesn't look like they had a cached installer, so the script downloaded 13.1 sparse image.

[check_free_space] OK - 67 GB free/purgeable disk space detected

67 GB of free space at the time the script was ran.

The error happens after the sparseimage is installed;

```
Product downloaded and installed to /Library/Management/erase-install/Install_macOS_13.1-22C65.sparseimage
   [run_installinstallmacos] Mounting sparse disk image to identify installer app.
   [erase-install] Installer is at: /Volumes/Install_macOS_13.1-22C65/Applications/Install macOS Ventura.app
   [erase-install] Closing DEPNotify download message (language=en)
/Library/Management/erase-install/erase-install.sh: line 1266: kill: (8614) - No such process
   [check_free_space] ERROR - 38 GB free/purgeable disk space detected
0
   [erase-install] attempting to terminate the 'caffeinate' process - Termination message indicates success
/Library/Management/erase-install/erase-install.sh: line 1046:  8340 Terminated: 15          /usr/bin/caffeinate -dimsu -w $$

   [finish] Script exit code: 1
```

It seems like it's erroring here:

`[check_free_space] ERROR - 38 GB free/purgeable disk space detected`

So the user ran it two more times.

Second time:

```
Running command /Library/Management/erase-install/erase-install.sh --reinstall --current-user --depnotify --cleanup-after-use --check-power...
--
Result of command:[erase-install] v27.1 script execution started: Thu Dec 15 14:36:23 PST 2022    [erase-install] Caffeinating this script (pid=11371)    [check_free_space] ERROR - 42 GB free/purgeable disk space detected 0    [erase-install] attempting to terminate the 'caffeinate' process - Termination message indicates success /Library/Management/erase-install/erase-install.sh: line 1046: 11384 Terminated: 15          /usr/bin/caffeinate -dimsu -w $     [finish] Script exit code: 1
```

Third time (this time it worked, but Jamf still showed 'Completed'):

```
Result of command:
[erase-install] v27.1 script execution started: Thu Dec 15 14:38:04 PST 2022
   [erase-install] Caffeinating this script (pid=12316)
   [check_free_space] ERROR - 42 GB free/purgeable disk space detected
0
   [erase-install] attempting to terminate the 'caffeinate' process - Termination message indicates success
/Library/Management/erase-install/erase-install.sh: line 1046: 12329 Terminated: 15          /usr/bin/caffeinate -dimsu -w $$

   [finish] Script exit code: 1
```

I've always read that if the script shows 'Completed' in Jamf policy logs, it didn't work.

I remoted into the computer after the third time to look it over, expecting it to still be on Monterey, but to my surprise, the update worked.

Info:

Cloud-hosted Jamf Pro 10.42.1-t1667311080
M1 MacBook Air
Was running Monterey 12.2.1
erase-install-depnotify-27.1.pkg
Parameters: 
/Library/Management/erase-install/erase-install.sh --reinstall --current-user --depnotify --cleanup-after-use --check-power

We're wanting to use an already cached Ventura installer in /Applications/ if it exists, and download from Apple if it does not.

TL:DR - The script did seem to work, it just took a couple tries. It also showed successful each time, which I didn't think was expected if it failed. And I wanted to ask about the check_free_space errors.

Thanks.
![Screenshot 2022-12-15 at 4 07 16 PM](https://user-images.githubusercontent.com/87879319/207992541-e3eac385-ce10-408f-98bb-953821ba318f.png)
![Screenshot 2022-12-15 at 4 07 06 PM](https://user-images.githubusercontent.com/87879319/207992544-2b7a4d1c-910c-4934-8dca-4f5efa986d5b.png)
![Screenshot 2022-12-15 at 4 06 54 PM](https://user-images.githubusercontent.com/87879319/207992546-4fa99403-cf44-4831-a2af-c6a4933524cb.png)

**Comment by itsbradwhite** [comment]
I have another user also getting the [check_free_space] ERROR. This user has a cached 'Install macOS Ventura.app' in /Applications/. It doesn't seem like erase-install is finding it.

I was testing the workflow of a non-admin end-user tries to run the Ventura update through Software Update, downloads the installer to /Applications/, realizes they're not able to run it... then goes to Self Service and runs erase-install there, which should be looking for a cached installer before kicking off a download of one.

Same parameters and everything as above, only difference is the computer:

M1 MacBook Air
Big Sur 11.7.1

```
[STEP 1 of 6]
--
Executing Policy Update to macOS Ventura
[STEP 2 of 6]
Running script Recon (Before)...
Script exit code: 0
Script result: Retrieving inventory preferences from https://jss.psd401.net:8443/...Finding extension attributes... Locating applications... Searching path: /System/Applications Locating hard drive information... Locating package receipts... Locating accounts... Gathering application usage information from the JamfDaemon... Searching path: /Applications Locating hardware information (macOS 11.7.1)... Submitting data to https://jss.psd401.net:8443/... <computer_id>3119</computer_id>
[STEP 3 of 6]
Downloading erase-install-depnotify-27.1.pkg...
Downloading https://use1-jcds.services.jamfcloud.com//download/b4d1560653744edfa928ef8710e48005/erase-install-depnotify-27.1.pkg?token=b154c444b96249c3b370cd12913a79a4css19t859t0sajzo916pvtyjmhfuhuxj...
Installing erase-install-depnotify-27.1.pkg...
Successfully installed erase-install-depnotify-27.1.pkg.
[STEP 4 of 6]
[STEP 5 of 6]
Running command /Library/Management/erase-install/erase-install.sh --reinstall --current-user --depnotify --cleanup-after-use --check-power...
Result of command:[erase-install] v27.1 script execution started: Fri Dec 16 10:23:39 PST 2022    [erase-install] Caffeinating this script (pid=62516)    [check_free_space] ERROR - 40 GB free/purgeable disk space detected 0    [erase-install] attempting to terminate the 'caffeinate' process - Termination message indicates success /Library/Management/erase-install/erase-install.sh: line 1046: 62530 Terminated: 15          /usr/bin/caffeinate -dimsu -w $     [finish] Script exit code: 1
[STEP 6 of 6]

```


they ran it again, space error differed slightly, but overall probably same:

```
[STEP 1 of 6]
--
Executing Policy Update to macOS Ventura
[STEP 2 of 6]
Running script Recon (Before)...
Script exit code: 0
Script result: Retrieving inventory preferences from https://jss.psd401.net:8443/...Finding extension attributes... Locating accounts... Locating applications... Searching path: /System/Applications Locating hard drive information... Locating package receipts... Gathering application usage information from the JamfDaemon... Searching path: /Applications Locating hardware information (macOS 11.7.1)... Submitting data to https://jss.psd401.net:8443/... <computer_id>3119</computer_id>
[STEP 3 of 6]
Downloading erase-install-depnotify-27.1.pkg...
Downloading https://use1-jcds.services.jamfcloud.com//download/b4d1560653744edfa928ef8710e48005/erase-install-depnotify-27.1.pkg?token=b154c444b96249c3b370cd12913a79a4css19t859t0sajzo916pvtyjmhfuhuxj...
Installing erase-install-depnotify-27.1.pkg...
Successfully installed erase-install-depnotify-27.1.pkg.
[STEP 4 of 6]
[STEP 5 of 6]
Running command /Library/Management/erase-install/erase-install.sh --reinstall --current-user --depnotify --cleanup-after-use --check-power...
Result of command:[erase-install] v27.1 script execution started: Fri Dec 16 11:41:06 PST 2022    [erase-install] Caffeinating this script (pid=68760)    [check_free_space] ERROR - 36 GB free/purgeable disk space detected 0    [erase-install] attempting to terminate the 'caffeinate' process - Termination message indicates success /Library/Management/erase-install/erase-install.sh: line 1046: 68773 Terminated: 15          /usr/bin/caffeinate -dimsu -w $     [finish] Script exit code: 1
[STEP 6 of 6]
```

Same parameters as above:

```
/Library/Management/erase-install/erase-install.sh --reinstall --current-user --depnotify --cleanup-after-use --check-power
```


![ventura_app_installer](https://user-images.githubusercontent.com/87879319/208179697-8fc55432-3a92-4f8d-a805-ed37cbf64d0a
![Screenshot 2022-12-16 at 12 03 46 PM](https://user-images.githubusercontent.com/87879319/208179707-fcb7fda7-39b7-4d8d-b2f8-be42f2041924.png)
![Screenshot 2022-12-16 at 12 03 37 PM](https://user-images.githubusercontent.com/87879319/208179709-b39eb3e0-cd5a-48cd-90ec-739259aae935.png)
.png)

**Comment by itsbradwhite** [comment]
![Screenshot 2022-12-16 at 2 15 55 PM](https://user-images.githubusercontent.com/87879319/208200283-b4a7a52c-0cc1-4c8d-be7b-ddd044857c4f.png)

**Comment by itsbradwhite** [comment]
![erase-install](https://user-images.githubusercontent.com/87879319/208200310-3bf65996-04f1-4d33-a70f-84d07b992a52.png)

**Comment by itsbradwhite** [comment]
erase-install-27.2.pkg

`/Library/Management/erase-install/erase-install.sh --build=22C65 --update --reinstall --current-user --depnotify --cleanup-after-use --check-power --no-timeout`

I've also tried

`/Library/Management/erase-install/erase-install.sh --reinstall --current-user --depnotify --cleanup-after-use --check-power`

Just can't get it to use the cached existing installer.

**Comment by grahampugh** [comment]
If you run the script from a run command it will always return successful. That's how Jamf works. If you want it to return a fail you need to use a launcher script. See the wiki for how to use the supplied launcher script.

**Comment by itsbradwhite** [comment]
I see. I thought I recalled reading in the past that it was normal for erase-install to stay pending in the logs, even after a successful run. I'm pretty sure that's how it used to work for us if I remember right. In any case, that's the last of my worries, it was more of an aside.

**Comment by grahampugh** [comment]
36 or 40 GB is not enough space. The limit is set at 45 GB. If you want to try your luck with a smaller amount of free space, you can override it (I'm on my phone so can't remember the flag to use - just run --help and it will tell you).

**Comment by grahampugh** [comment]
It will stay pending if it successfully restarts the computer, because the policy would then never complete. But if the command fails, the policy completes so it ships the logs.

**Comment by itsbradwhite** [comment]
Understood. Lots of places are saying the Ventura installer requires ~35GB as opposed to Monterey's ~45GB, so that's good to know.

So does having a cached installer not really help with the space issue? It seems like if anything, it's exasperating the issue then. I guess we'll have to use the --overwrite flag.

**Comment by grahampugh** [comment]
Or just override the`--min-drive-space` flag with a value you think might work, and test it. I don't have any clever logic in that section to work out what is really required, so it is a relatively conservative value.

**Comment by itsbradwhite** [comment]
Understood. Let's disregard any issues where the script is starting with less than 45GB of free space.

```
[STEP 1 of 6]
--
Executing Policy Update to macOS Ventura
[STEP 2 of 6]
Running script Delete Cached macOS Installers from Applications...
Script exit code: 0
Script result: /Applications/Install\ macOS\ Ventura.app does not on your filesystem, moving on.../Applications/Install\ macOS\ Monterey.app does not on your filesystem, moving on... /Applications/Install\ macOS\ Big Sur.app does not on your filesystem, moving on...
[STEP 3 of 6]
Downloading erase-install-27.2.pkg...
Downloading https://use1-jcds.services.jamfcloud.com//download/b4d1560653744edfa928ef8710e48005/erase-install-27.2.pkg?token=34f1c10e55d042c989a220fa3567a2b84nn6pd3301939ka9z7aszh5ame1a6f72...
Installing erase-install-27.2.pkg...
Successfully installed erase-install-27.2.pkg.
[STEP 4 of 6]
[STEP 5 of 6]
Running command /Library/Management/erase-install/erase-install.sh --overwrite --reinstall --current-user --depnotify --cleanup-after-use --check-power --no-timeout...
Result of command:[erase-install] v27.2 script execution started: Tue Dec 20 11:42:02 PST 2022    [erase-install] Caffeinating this script (pid=2878)    [check_free_space] OK - 50 GB free/purgeable disk space detected    [check_power_status] OK - AC power detected    [erase-install] Looking for existing installer app or pkg    [find_existing_installer] No valid installer found.    [overwrite_existing_installer] Deleting existing cached installer(s).    [erase-install] Running on architecture arm64 yes squiresd is a member of everyone    [get_user_details] squiresd is a Volume Owner    [get_user_details] ask for password (attempt 1/5)    [check_password] Success: the password entered is the correct login password for squiresd.    [erase-install] Opening DEPNotify download message (language=en)    [check_installinstallmacos] installinstallmacos.py is in /Library/Management/erase-install    [check_python] Relocatable Python is installed in /Library/Management/erase-install    [run_installinstallmacos] Getting current production version     [run_installinstallmacos] This command is now being run:     installinstallmacos.py --workdir /Library/Management/erase-install --ignore-cache --raw --auto   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                  Dload  Upload   Total   Spent    Left  Speed    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0  12 3379k   12  419k    0     0  1266k      0  0:00:02 --:--:--  0:00:02 1266k 100 3379k  100 3379k    0     0  2767k      0  0:00:01  0:00:01 --:--:-- 2767k   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                  Dload  Upload   Total   Spent    Left  Speed    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0   0 3276k    0  6093    0     0  63468      0  0:00:52 --:--:--  0:00:52 62814  50 3276k   50 1662k    0     0  1521k      0  0:00:02  0:00:01  0:00:01 1520k 100 3276k  100 3276k    0     0  1795k      0  0:00:01  0:00:01 --:--:-- 1794k   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                  Dload  Upload   Total   Spent    Left  Speed    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0   0 11.3G    0 13282    0     0  17972      0   7d 20h --:--:--   7d 20h 17948   0 11.3G    0 2030k    0     0  1620k      0  2:02:41  0:00:01  2:02:40 1619k   0 11.3G    0 5756k    0     0  2556k      0  1:17:47  0:00:02  1:17:45 2555k   0 11.3G    0 7933k    0     0  2382k      0  1:23:27  0:00:03  1:23:24 2381k   0 11.3G    
```

[ I cut-out a lot of the unnecessary download part ]

```
1:26:52  0:00:13 2161k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:05  1:26:53  0:00:12 2111k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:05  1:26:54  0:00:11 2431k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:05  1:26:55  0:00:10 2532k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:05  1:26:56  0:00:09 2326k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:05  1:26:57  0:00:08 2329k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:05  1:26:58  0:00:07 2356k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:05  1:26:59  0:00:06 2324k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:05  1:27:00  0:00:05 2238k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:05  1:27:01  0:00:04 2431k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:04  1:27:02  0:00:02 2482k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:04  1:27:03  0:00:01 2526k  99 11.3G   99 11.3G    0     0  2283k      0  1:27:04  1:27:04 --:--:-- 2346k 100 11.3G  100 11.3G    0     0  2283k      0  1:27:04  1:27:04 --:--:-- 2494k   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                  Dload  Upload   Total   Spent    Left  Speed    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0 100 1303k  100 1303k    0     0  1900k      0 --:--:-- --:--:-- --:--:-- 1900k   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                  Dload  Upload   Total   Spent    Left  Speed    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0 100  5041  100  5041    0     0  77553      0 --:--:-- --:--:-- --:--:-- 77553   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                  Dload  Upload   Total   Spent    Left  Speed    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0 100   181  100   181    0     0    760      0 --:--:-- --:--:-- --:--:--   760 installer: Package name is macOS Ventura installer: Installing at base path /private/tmp/dmg.yMI1ER installer: The install was successful.  installinstallmacos.py - get macOS installers from the Apple software catalog  This Mac: Model Identifier : MacBookAir10,1  Device ID        : J313AP OS Version       : 11.7.1 Build ID         : 20G918   #  ProductID       Version    Build    Post Date   Title                            1  061-26578       10.14.5    18F2059  2019-10-14  macOS Mojave                     2  061-26589       10.14.6    18G103   2019-10-14  macOS Mojave                     3  041-91758       10.13.6    17G66    2019-10-19  macOS High Sierra                4  041-88800       10.14.4    18E2034  2019-10-23  macOS Mojave                     5  041-90855       10.13.5    17F66a   2019-10-23  Install macOS High Sierra Beta   6  061-86291       10.15.3    19D2064  2020-03-23  macOS Catalina                   7  001-04366       10.15.4    19E2269  2020-05-04  macOS Catalina                   8  001-15219       10.15.5    19F2200  2020-06-15  macOS Catalina                   9  001-36735       10.15.6    19G2006  2020-08-06  macOS Catalina                  10  001-36801       10.15.6    19G2021  2020-08-12  macOS Catalina                  11  001-51042       10.15.7    19H2     2020-09-24  macOS Catalina                  12  001-57224       10.15.7    19H4     2020-10-27  macOS Catalina                  13  001-68446       10.15.7    19H15    2020-11-11  macOS Catalina                  14  071-78704       11.5.2     20G95    2021-08-18  macOS Big Sur                   15  002-23589       11.6.1     20G224   2021-12-01  macOS Big Sur                   16  002-42341       11.6.2     20G314   2022-01-14  macOS Big Sur                   17  002-57023       11.6.3     20G415   2022-01-26  macOS Big Sur                   18  002-65695       11.6.4     20G417   2022-02-17  macOS Big Sur                   19  002-77154       11.6.5     20G527   2022-04-11  macOS Big Sur                   20  012-08272       11.6.6     20G624   2022-05-24  macOS Big Sur                   21  012-40494       12.6       21G115   2022-09-20  macOS Monterey                  22  012-38280       11.7       20G817   2022-09-20  macOS Big Sur                   23  012-92138       13.0       22A380   2022-10-24  macOS Ventura                   24  012-90254       12.6.1     21G217   2022-10-24  macOS Monterey                  25  012-90253       11.7.1     20G918   2022-10-24  macOS Big Sur                   26  012-93766       13.0.1     22A400   2022-11-09  macOS Ventura                   27  012-60271       13.1       22C65    2022-12-13  macOS Ventura                   28  032-12843       12.6.2     21G320   2022-12-13  macOS Monterey                  29  032-12833       11.7.2     20G1020  2022-12-13  macOS Big Sur                    Build 22C65 selected. Downloading #27...  Making empty sparseimage... ********************************************************* *** Working around a very dumb Apple bug in a package *** *** postinstall script that fails to correctly target *** *** the Install macOS.app when installed to a volume  *** *** other than the current boot volume.               *** ***       Please file feedback with Apple!            *** ********************************************************* Product downloaded and installed to /Library/Management/erase-install/Install_macOS_13.1-22C65.sparseimage    [run_installinstallmacos] Mounting sparse disk image to identify installer app.    [erase-install] Installer is at: /Volumes/Install_macOS_13.1-22C65/Applications/Install macOS Ventura.app    [erase-install] Closing DEPNotify download message (language=en) /Library/Management/erase-install/erase-install.sh: line 1266: kill: (2958) - No such process    [check_free_space] ERROR - 36 GB free/purgeable disk space detected 0    [erase-install] attempting to terminate the 'caffeinate' process - Termination message indicates success /Library/Management/erase-install/erase-install.sh: line 1046:  2891 Terminated: 15          /usr/bin/caffeinate -dimsu -w $     [finish] Script exit code: 1
[STEP 6 of 6]
```

Script said OK - 50 GB free, kicked off the Ventura 13.1 sparse image download, then errors after the second space check.

Is that expected? Or is the 45GB still needed _after_ the sparse image file is downloaded?

Computer is on 11.7.1
erase-install-27.2.pkg
Flags: --overwrite --reinstall --current-user --depnotify --cleanup-after-use --check-power --no-timeout

**Comment by grahampugh** [comment]
I don't know the exact figures or whether the required space includes the installer or not - I don't think Apple publish exact figures. But, people were having issues with failed installs when the space check was not done again after downloading, so we added that in. 

So, yes, if you don't have 45 GB spare at the point of install it is currently expected that it will bail.

**Comment by itsbradwhite** [comment]
I really appreciate you taking the time to respond. I guess that solves all of my questions and it comes down to a space issue and not anything wrong with the script.
